From 864d70f4052a84c5dbd4a386fde14bea2a260467 Mon Sep 17 00:00:00 2001
From: Bai Ping <ping.bai@nxp.com>
Date: Mon, 21 Nov 2016 14:56:12 +0800
Subject: [PATCH 1362/5242] MLK-13616 ARM: imx: Add low power run voltage
 change support on i.MX6SLL

commit  95b660e13e5d2f524aaeb9980fdb1f2b61bf8bea from
https://source.codeaurora.org/external/imx/linux-imx.git

On i.MX6SLL, if all PLLs is bypassed in low power run mode, we can decrease
the VDD_ARM_IN and VDD_SOC_IN voltage to 0.925V to save power. a 25mV margin
is added to cover IR drop and board tolerance.

Add low power run voltage change support for i.MX6SLL.

Signed-off-by: Bai Ping <ping.bai@nxp.com>
Signed-off-by: Meng Li <Meng.Li@windriver.com>
---
 arch/arm/mach-imx/busfreq-imx.c |   28 +++++++++++++++++++++++++++-
 arch/arm/mach-imx/common.h      |    2 ++
 drivers/cpufreq/imx6q-cpufreq.c |    4 ++--
 include/linux/busfreq-imx.h     |    5 ++++-
 4 files changed, 35 insertions(+), 4 deletions(-)

diff --git a/arch/arm/mach-imx/busfreq-imx.c b/arch/arm/mach-imx/busfreq-imx.c
index 71ee54f..cc8da8c 100644
--- a/arch/arm/mach-imx/busfreq-imx.c
+++ b/arch/arm/mach-imx/busfreq-imx.c
@@ -39,6 +39,8 @@
 #define LOW_AUDIO_CLK		50000000
 #define HIGH_AUDIO_CLK		100000000
 
+#define LOW_POWER_RUN_VOLTAGE	950000
+
 #define MMDC_MDMISC_DDR_TYPE_DDR3	0
 #define MMDC_MDMISC_DDR_TYPE_LPDDR2	1
 
@@ -62,6 +64,7 @@
 static unsigned int ddr_low_rate;
 static int cur_bus_freq_mode;
 static u32 org_arm_rate;
+static int origin_arm_volt, origin_soc_volt;
 
 extern unsigned long iram_tlb_phys_addr;
 extern int unsigned long iram_tlb_base_addr;
@@ -164,8 +167,15 @@ int unregister_busfreq_notifier(struct notifier_block *nb)
  */
 static void imx6ull_lower_cpu_rate(bool enter)
 {
-	if (enter)
+	int ret;
+
+	if (enter) {
 		org_arm_rate = clk_get_rate(arm_clk);
+		if (cpu_is_imx6sll()) {
+			origin_arm_volt = regulator_get_voltage(arm_reg);
+			origin_soc_volt = regulator_get_voltage(soc_reg);
+		}
+	}
 
 	clk_set_parent(pll1_bypass_clk, pll1_bypass_src_clk);
 	clk_set_parent(pll1_sw_clk, pll1_sys_clk);
@@ -175,7 +185,23 @@ static void imx6ull_lower_cpu_rate(bool enter)
 		clk_set_parent(step_clk, osc_clk);
 		clk_set_parent(pll1_sw_clk, step_clk);
 		clk_set_rate(arm_clk, LPAPM_CLK);
+		if (cpu_is_imx6sll() && uart_from_osc) {
+			ret = regulator_set_voltage_tol(arm_reg, LOW_POWER_RUN_VOLTAGE, 0);
+			if (ret)
+				pr_err("set arm reg voltage failed\n");
+			ret = regulator_set_voltage_tol(soc_reg, LOW_POWER_RUN_VOLTAGE, 0);
+			if (ret)
+				pr_err("set soc reg voltage failed\n");
+		}
 	} else {
+		if (cpu_is_imx6sll() && uart_from_osc) {
+			ret = regulator_set_voltage_tol(soc_reg, origin_soc_volt, 0);
+			if (ret)
+				pr_err("set soc reg voltage failed\n");
+			ret = regulator_set_voltage_tol(arm_reg, origin_arm_volt, 0);
+			if (ret)
+				pr_err("set arm reg voltage failed\n");
+		}
 		clk_set_parent(step_clk, origin_step_parent);
 		clk_set_parent(pll1_sw_clk, step_clk);
 		clk_set_rate(arm_clk, org_arm_rate);
diff --git a/arch/arm/mach-imx/common.h b/arch/arm/mach-imx/common.h
index 1b52de6..9ddc782 100644
--- a/arch/arm/mach-imx/common.h
+++ b/arch/arm/mach-imx/common.h
@@ -234,4 +234,6 @@ static inline void imx_init_l2cache(void) {}
 extern const struct smp_operations imx_smp_ops;
 extern const struct smp_operations ls1021a_smp_ops;
 
+extern bool uart_from_osc;
+
 #endif
diff --git a/drivers/cpufreq/imx6q-cpufreq.c b/drivers/cpufreq/imx6q-cpufreq.c
index 19dd599..01c13a2 100644
--- a/drivers/cpufreq/imx6q-cpufreq.c
+++ b/drivers/cpufreq/imx6q-cpufreq.c
@@ -29,9 +29,9 @@
 #define FREQ_198_MHZ		198000
 #define FREQ_24_MHZ		24000
 
-static struct regulator *arm_reg;
+struct regulator *arm_reg;
 static struct regulator *pu_reg;
-static struct regulator *soc_reg;
+struct regulator *soc_reg;
 static struct regulator *dc_reg;
 
 enum IMX6_CPUFREQ_CLKS {
diff --git a/include/linux/busfreq-imx.h b/include/linux/busfreq-imx.h
index 35d57fe..6c98244 100644
--- a/include/linux/busfreq-imx.h
+++ b/include/linux/busfreq-imx.h
@@ -1,5 +1,5 @@
 /*
- * Copyright 2012-2015 Freescale Semiconductor, Inc. All Rights Reserved.
+ * Copyright 2012-2016 Freescale Semiconductor, Inc. All Rights Reserved.
  *
  * This program is free software; you can redistribute it and/or modify
  * it under the terms of the GNU General Public License version 2 as
@@ -10,6 +10,7 @@
 #define __ASM_ARCH_MXC_BUSFREQ_H__
 
 #include <linux/notifier.h>
+#include <linux/regulator/consumer.h>
 
 /*
  * This enumerates busfreq low power mode entry and exit.
@@ -42,6 +43,8 @@ enum bus_freq_mode {
 };
 
 #if defined(CONFIG_CPU_FREQ) && !defined(CONFIG_ARM64)
+extern struct regulator *arm_reg;
+extern struct regulator *soc_reg;
 void request_bus_freq(enum bus_freq_mode mode);
 void release_bus_freq(enum bus_freq_mode mode);
 int register_busfreq_notifier(struct notifier_block *nb);
-- 
1.7.9.5

